Stacktrace should tollerate memory read errors

Occasionally some of the tests will fail with a memory read error while doing a stack trace. [This is an example with the gdbserial backend on macOS](https://travis-ci.org/derekparker/delve/jobs/241083651):

```
=== RUN   TestScopePrefix
--- FAIL: TestScopePrefix (0.36s)
	support.go:129: enabling recording for github.com/derekparker/delve/pkg/terminal.TestScopePrefix
	command_test.go:73: Error executing <goroutine 8 stack>: protocol error E08 during memory read for packet $m28,8
```

but I've also seen it happen with the linux native backend. The likely cause is the:

```
	ret, err := readUintRaw(it.mem, retaddr, int64(it.bi.Arch.PtrSize()))
	if err != nil {
		return Stackframe{}, err
	}
```

in pkg/proc/stack.go. We should tolerate memory read errors during stack traces and return partial stack traces instead.

--- a/pkg/proc/gdbserial/gdbserver.go
+++ b/pkg/proc/gdbserial/gdbserver.go
@@ -623,25 +623,26 @@ continueLoop:
 }
 
 func (p *Process) StepInstruction() error {
-	if p.selectedGoroutine == nil {
-		return errors.New("cannot single step: no selected goroutine")
-	}
-	if p.selectedGoroutine.Thread == nil {
-		if _, err := p.SetBreakpoint(p.selectedGoroutine.PC, proc.NextBreakpoint, proc.SameGoroutineCondition(p.selectedGoroutine)); err != nil {
-			return err
+	thread := p.currentThread
+	if p.selectedGoroutine != nil {
+		if p.selectedGoroutine.Thread == nil {
+			if _, err := p.SetBreakpoint(p.selectedGoroutine.PC, proc.NextBreakpoint, proc.SameGoroutineCondition(p.selectedGoroutine)); err != nil {
+				return err
+			}
+			return proc.Continue(p)
 		}
-		return proc.Continue(p)
+		thread = p.selectedGoroutine.Thread.(*Thread)
 	}
 	p.allGCache = nil
 	if p.exited {
 		return &proc.ProcessExitedError{Pid: p.conn.pid}
 	}
-	p.selectedGoroutine.Thread.(*Thread).clearBreakpointState()
-	err := p.selectedGoroutine.Thread.(*Thread).StepInstruction()
+	thread.clearBreakpointState()
+	err := thread.StepInstruction()
 	if err != nil {
 		return err
 	}
-	return p.selectedGoroutine.Thread.(*Thread).SetCurrentBreakpoint()
+	return thread.SetCurrentBreakpoint()
 }
 
 func (p *Process) SwitchThread(tid int) error {

--- a/pkg/proc/gdbserial/gdbserver.go
+++ b/pkg/proc/gdbserial/gdbserver.go
@@ -642,7 +642,14 @@ func (p *Process) StepInstruction() error {
 	if err != nil {
 		return err
 	}
-	return thread.SetCurrentBreakpoint()
+	err = thread.SetCurrentBreakpoint()
+	if err != nil {
+		return err
+	}
+	if g, _ := proc.GetG(thread); g != nil {
+		p.selectedGoroutine = g
+	}
+	return nil
 }
 
 func (p *Process) SwitchThread(tid int) error {

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -1,7 +1,6 @@
 package native
 
 import (
-	"errors"
 	"fmt"
 	"go/ast"
 	"os"
@@ -297,26 +296,27 @@ func (dbp *Process) ContinueOnce() (proc.Thread, error) {
 // asssociated with the selected goroutine. All other
 // threads will remain stopped.
 func (dbp *Process) StepInstruction() (err error) {
-	if dbp.selectedGoroutine == nil {
-		return errors.New("cannot single step: no selected goroutine")
-	}
-	if dbp.selectedGoroutine.Thread == nil {
-		// Step called on parked goroutine
-		if _, err := dbp.SetBreakpoint(dbp.selectedGoroutine.PC, proc.NextBreakpoint, proc.SameGoroutineCondition(dbp.selectedGoroutine)); err != nil {
-			return err
+	thread := dbp.currentThread
+	if dbp.selectedGoroutine != nil {
+		if dbp.selectedGoroutine.Thread == nil {
+			// Step called on parked goroutine
+			if _, err := dbp.SetBreakpoint(dbp.selectedGoroutine.PC, proc.NextBreakpoint, proc.SameGoroutineCondition(dbp.selectedGoroutine)); err != nil {
+				return err
+			}
+			return proc.Continue(dbp)
 		}
-		return proc.Continue(dbp)
+		thread = dbp.selectedGoroutine.Thread.(*Thread)
 	}
 	dbp.allGCache = nil
 	if dbp.exited {
 		return &proc.ProcessExitedError{}
 	}
-	dbp.selectedGoroutine.Thread.(*Thread).clearBreakpointState()
-	err = dbp.selectedGoroutine.Thread.(*Thread).StepInstruction()
+	thread.clearBreakpointState()
+	err = thread.StepInstruction()
 	if err != nil {
 		return err
 	}
-	return dbp.selectedGoroutine.Thread.(*Thread).SetCurrentBreakpoint()
+	return thread.SetCurrentBreakpoint()
 }
 
 // SwitchThread changes from current thread to the thread specified by `tid`.

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -316,7 +316,14 @@ func (dbp *Process) StepInstruction() (err error) {
 	if err != nil {
 		return err
 	}
-	return thread.SetCurrentBreakpoint()
+	err = thread.SetCurrentBreakpoint()
+	if err != nil {
+		return err
+	}
+	if g, _ := proc.GetG(thread); g != nil {
+		dbp.selectedGoroutine = g
+	}
+	return nil
 }
 
 // SwitchThread changes from current thread to the thread specified by `tid`.

--- a/pkg/proc/proc.go
+++ b/pkg/proc/proc.go
@@ -80,12 +80,8 @@ func Next(dbp Process) (err error) {
 	}
 
 	if err = next(dbp, false); err != nil {
-		switch err.(type) {
-		case ThreadBlockedError, NoReturnAddr: // Noop
-		default:
-			dbp.ClearInternalBreakpoints()
-			return
-		}
+		dbp.ClearInternalBreakpoints()
+		return
 	}
 
 	return Continue(dbp)

--- a/pkg/proc/proc_test.go
+++ b/pkg/proc/proc_test.go
@@ -19,6 +19,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/derekparker/delve/pkg/dwarf/frame"
 	"github.com/derekparker/delve/pkg/proc"
 	"github.com/derekparker/delve/pkg/proc/gdbserial"
 	"github.com/derekparker/delve/pkg/proc/native"
@@ -2952,5 +2953,31 @@ func TestIssue877(t *testing.T) {
 		if vv != envval {
 			t.Fatalf("value of v is %q (expected %q)", vv, envval)
 		}
+	})
+}
+
+func TestIssue893(t *testing.T) {
+	// Test what happens when next is called immediately after launching the
+	// executable, acceptable behaviors are: (a) no error, (b) no source at PC
+	// error.
+	protest.AllowRecording(t)
+	withTestProcess("increment", t, func(p proc.Process, fixture protest.Fixture) {
+		err := proc.Next(p)
+		if err == nil {
+			return
+		}
+		if _, ok := err.(*frame.NoFDEForPCError); ok {
+			return
+		}
+		assertNoError(err, t, "Next")
+	})
+}
+
+func TestStepInstructionNoGoroutine(t *testing.T) {
+	protest.AllowRecording(t)
+	withTestProcess("increment", t, func(p proc.Process, fixture protest.Fixture) {
+		// Call StepInstruction immediately after launching the program, it should
+		// work even though no goroutine is selected.
+		assertNoError(p.StepInstruction(), t, "StepInstruction")
 	})
 }

--- a/pkg/proc/proc_test.go
+++ b/pkg/proc/proc_test.go
@@ -2969,6 +2969,9 @@ func TestIssue893(t *testing.T) {
 		if _, ok := err.(*frame.NoFDEForPCError); ok {
 			return
 		}
+		if _, ok := err.(proc.ThreadBlockedError); ok {
+			return
+		}
 		assertNoError(err, t, "Next")
 	})
 }

--- a/pkg/proc/stack.go
+++ b/pkg/proc/stack.go
@@ -38,6 +38,8 @@ type Stackframe struct {
 	Ret uint64
 	// Address to the memory location containing the return address
 	addrret uint64
+	// Err is set if an error occoured during stacktrace
+	Err error
 }
 
 // Stacktrace returns the stack trace for thread.
@@ -207,7 +209,7 @@ func (it *stackIterator) newStackframe(pc uint64, cfa int64, retaddr uintptr, fd
 	f, l, fn := it.bi.PCToLine(pc)
 	ret, err := readUintRaw(it.mem, retaddr, int64(it.bi.Arch.PtrSize()))
 	if err != nil {
-		return Stackframe{}, err
+		it.err = err
 	}
 	r := Stackframe{Current: Location{PC: pc, File: f, Line: l, Fn: fn}, CFA: cfa, FDE: fde, Ret: ret, addrret: uint64(retaddr), StackHi: it.stackhi}
 	if !top {
@@ -231,7 +233,10 @@ func (it *stackIterator) stacktrace(depth int) ([]Stackframe, error) {
 		}
 	}
 	if err := it.Err(); err != nil {
-		return nil, err
+		if len(frames) == 0 {
+			return nil, err
+		}
+		frames = append(frames, Stackframe{Err: err})
 	}
 	return frames, nil
 }

--- a/pkg/proc/threads.go
+++ b/pkg/proc/threads.go
@@ -407,7 +407,7 @@ func onRuntimeBreakpoint(thread Thread) bool {
 func onNextGoroutine(thread Thread, breakpoints map[uint64]*Breakpoint) (bool, error) {
 	var bp *Breakpoint
 	for i := range breakpoints {
-		if breakpoints[i].Internal() {
+		if breakpoints[i].Internal() && breakpoints[i].Cond != nil {
 			bp = breakpoints[i]
 			break
 		}

--- a/pkg/proc/threads.go
+++ b/pkg/proc/threads.go
@@ -112,7 +112,8 @@ func next(dbp Process, stepInto bool) error {
 		}
 	}()
 
-	csource := filepath.Ext(topframe.Current.File) != ".go"
+	ext := filepath.Ext(topframe.Current.File)
+	csource := ext != ".go" && ext != ".s"
 	var thread MemoryReadWriter = curthread
 	var regs Registers
 	if selg != nil && selg.Thread != nil {

--- a/pkg/proc/threads.go
+++ b/pkg/proc/threads.go
@@ -4,7 +4,6 @@ import (
 	"debug/gosym"
 	"encoding/binary"
 	"errors"
-	"fmt"
 	"go/ast"
 	"go/token"
 	"path/filepath"
@@ -127,12 +126,6 @@ func next(dbp Process, stepInto bool) error {
 	text, err := disassemble(thread, regs, dbp.Breakpoints(), dbp.BinInfo(), topframe.FDE.Begin(), topframe.FDE.End())
 	if err != nil && stepInto {
 		return err
-	}
-
-	for i := range text {
-		if text[i].Inst == nil {
-			fmt.Printf("error at instruction %d\n", i)
-		}
 	}
 
 	sameGCond := SameGoroutineCondition(selg)

--- a/pkg/proc/threads.go
+++ b/pkg/proc/threads.go
@@ -250,10 +250,11 @@ func next(dbp Process, stepInto bool) error {
 				// this frame and on the return frame.
 				bp.Cond = sameOrRetFrameCond
 			}
-		} else {
-			return err
 		}
+		// Return address could be wrong, if we are unable to set a breakpoint
+		// there it's ok.
 	}
+
 	if bp, _, _ := curthread.Breakpoint(); bp == nil {
 		curthread.SetCurrentBreakpoint()
 	}

--- a/pkg/terminal/command.go
+++ b/pkg/terminal/command.go
@@ -1210,6 +1210,10 @@ func printStack(stack []api.Stackframe, ind string) {
 	s := ind + strings.Repeat(" ", d+2+len(ind))
 
 	for i := range stack {
+		if stack[i].Err != "" {
+			fmt.Printf("%serror: %s\n", s, stack[i].Err)
+			continue
+		}
 		name := "(nil)"
 		if stack[i].Function != nil {
 			name = stack[i].Function.Name

--- a/pkg/terminal/command_test.go
+++ b/pkg/terminal/command_test.go
@@ -216,7 +216,7 @@ func TestExecuteFile(t *testing.T) {
 
 func TestIssue354(t *testing.T) {
 	printStack([]api.Stackframe{}, "")
-	printStack([]api.Stackframe{{api.Location{PC: 0, File: "irrelevant.go", Line: 10, Function: nil}, nil, nil, 0, nil}}, "")
+	printStack([]api.Stackframe{{api.Location{PC: 0, File: "irrelevant.go", Line: 10, Function: nil}, nil, nil, 0, ""}}, "")
 }
 
 func TestIssue411(t *testing.T) {

--- a/pkg/terminal/command_test.go
+++ b/pkg/terminal/command_test.go
@@ -216,7 +216,7 @@ func TestExecuteFile(t *testing.T) {
 
 func TestIssue354(t *testing.T) {
 	printStack([]api.Stackframe{}, "")
-	printStack([]api.Stackframe{{api.Location{PC: 0, File: "irrelevant.go", Line: 10, Function: nil}, nil, nil, 0}}, "")
+	printStack([]api.Stackframe{{api.Location{PC: 0, File: "irrelevant.go", Line: 10, Function: nil}, nil, nil, 0, nil}}, "")
 }
 
 func TestIssue411(t *testing.T) {

--- a/service/api/types.go
+++ b/service/api/types.go
@@ -121,6 +121,7 @@ type Stackframe struct {
 	Locals      []Variable
 	Arguments   []Variable
 	FrameOffset int64
+	Err         string
 }
 
 func (frame *Stackframe) Var(name string) *Variable {

--- a/service/debugger/debugger.go
+++ b/service/debugger/debugger.go
@@ -538,7 +538,7 @@ func (d *Debugger) Command(command *api.DebuggerCommand) (*api.DebuggerState, er
 	}
 
 	if err != nil {
-		if exitedErr, exited := err.(proc.ProcessExitedError); withBreakpointInfo && exited {
+		if exitedErr, exited := err.(proc.ProcessExitedError); (command.Name == api.Continue || command.Name == api.Rewind) && exited {
 			state := &api.DebuggerState{}
 			state.Exited = true
 			state.ExitStatus = exitedErr.Status

--- a/service/debugger/debugger.go
+++ b/service/debugger/debugger.go
@@ -857,6 +857,9 @@ func (d *Debugger) convertStacktrace(rawlocs []proc.Stackframe, cfg *proc.LoadCo
 			Location:    api.ConvertLocation(rawlocs[i].Call),
 			FrameOffset: rawlocs[i].CFA - int64(rawlocs[i].StackHi),
 		}
+		if rawlocs[i].Err != nil {
+			frame.Err = rawlocs[i].Err.Error()
+		}
 		if cfg != nil && rawlocs[i].Current.Fn != nil {
 			var err error
 			scope := proc.FrameToScope(d.target, rawlocs[i])

--- a/service/debugger/debugger.go
+++ b/service/debugger/debugger.go
@@ -594,6 +594,12 @@ func (d *Debugger) collectBreakpointInformation(state *api.DebuggerState) error 
 		if !found {
 			return fmt.Errorf("could not find thread %d", state.Threads[i].ID)
 		}
+
+		if len(bp.Variables) == 0 && bp.LoadArgs == nil && bp.LoadLocals == nil {
+			// don't try to create goroutine scope if there is nothing to load
+			continue
+		}
+
 		s, err := proc.GoroutineScope(thread)
 		if err != nil {
 			return err
@@ -605,9 +611,10 @@ func (d *Debugger) collectBreakpointInformation(state *api.DebuggerState) error 
 		for i := range bp.Variables {
 			v, err := s.EvalVariable(bp.Variables[i], proc.LoadConfig{true, 1, 64, 64, -1})
 			if err != nil {
-				return err
+				bpi.Variables[i] = api.Variable{Name: bp.Variables[i], Unreadable: fmt.Sprintf("eval error: %v", err)}
+			} else {
+				bpi.Variables[i] = *api.ConvertVar(v)
 			}
-			bpi.Variables[i] = *api.ConvertVar(v)
 		}
 		if bp.LoadArgs != nil {
 			if vars, err := s.FunctionArguments(*api.LoadConfigToProc(bp.LoadArgs)); err == nil {

--- a/service/test/integration2_test.go
+++ b/service/test/integration2_test.go
@@ -1371,3 +1371,16 @@ func TestClientServer_collectBreakpointInfoOnNext(t *testing.T) {
 		}
 	})
 }
+
+func TestClientServer_collectBreakpointInfoError(t *testing.T) {
+	protest.AllowRecording(t)
+	withTestClient2("testnextprog", t, func(c service.Client) {
+		_, err := c.CreateBreakpoint(&api.Breakpoint{
+			Addr:       findLocationHelper(t, c, "testnextprog.go:23", false, 1, 0)[0],
+			Variables:  []string{"nonexistentvariable", "j"},
+			LoadLocals: &normalLoadConfig})
+		assertNoError(err, t, "CreateBreakpoint()")
+		state := <-c.Continue()
+		assertNoError(state.Err, t, "Continue()")
+	})
+}

